### PR TITLE
fix(54152): Cannot use const enum to define an interface key in isolatedModules mode

### LINKED ISSUE #54152: Cannot use `const enum` to define an interface key in `isolatedModules` mode
# Bug Report

I am maintaning a library where I use `const enum` variants to declare the keys of an interface (`interface Foo { [MyConstEnum.Foo]: "bar" }`). The enum and the interface are in the same module.

It works well, but once compiled using `tsc` and included in a user project using `isolatedModules: true`, TypeScript fails with the following error:

>  Error: [...].d.ts: error TS2748: Cannot access ambient const enums when 'isolatedModules' is enabled.

I don't understand why this usage is restricted, since the enum is only used in a type definition, and both are present in the same `.d.ts` file.

### 🔎 Search Terms

Cannot access ambient const enums when the '--isolatedModules' flag is provided.
TS2748
interface

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about isolated modules and const enum

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?isolatedModules=true&ts=5.0.4#code/PTAEEkDtQewVwE6gDYEsBGAaUB3ApqACZ4DGyAhggeaCTJAM4AuoekcAtrqkwBZGoAZoLxVILPADc2LSOQ54G2Pm1DloAIikzQHcgAcNahgxglU5JqkgBzNdG3jQTAJ76CgmEjzkS-R7LyeAB0ALAAUCCgACq8MAwEcgoMalSgkDAs1kxsxITOMKDoBHAJ+eguzrwEaOgIlJWlogzBoAASonjY+ET0AOQS0g4AHvpeLCocYeERdIwS7FwAokNMAHJBoADeEaB7oABiAPJHoAC8oACMmLv7AEIAggBK56AATBEAvhERru6gKxkAFkDK8duF9qAANqA8QbBTBY5HAC6AC50pxiggANy3PYw1bwkKPJ5o0DMBDWGy48KfGkRKKxVApHBeADWKUE1i6RTgLHoJAITHqjH0qGQeHKlQGDBIfWU1VAAANgoRgkwGErQFyJbR6ExyNYUqYFHwqShUGyhbxmaifuFiGRKAQ5sxWIsAYSgm9tnjDidXtc-STXh9afa-gQAAoIGDoCV6KwkWFMEH6H0XcGQgkyIlvREnMmLLE07MpvPBElkilUml0+1RADq1Wg8CQTQQKVdnGtNQw9QQlWsVTwqCQ+ljACtSExsEIR5U5lybIgCNE3HgAMokSn6Fg4Hj8JXMmAUHKEIEwQhwCUMdHCuB4JUKjyn5AwA+2VgIWNIBhxHAUjgfQ7UiMA9gAYXUDIWF8QUTDUDh0FQHRXQWTgWRbUA+hPM9JUva9bz6UBmXdch40laYogATXgWh1HJPACB4NR0BgaRQAAPQbMA1g-bBwCKPA0G0KpSNEX8SJSaw6B-GdsAYGS8FAqI9gAKlwAhCH6WCSHgrt6DdNgMO6Ah6GQRoEhHMdnA3UAAAphxoJVIy1R0KHqKx6AASgZcDQA07RBzNL9SPc518mHFRyU2DgrxvPAeNAB4UhoVkEDZSh4EgQhsHCykvwY7lhmZA0nDQ7g+BHGKFFASRyGQR9UpSaLjK4BI8A4FrCnStlQIifKXUMlhy29FKIOG9Es32JF0VGhR8yRYNnjmr0FsrZ4vgjOzG3ZKkUzTDNfQhfYczhMaGAm+YCxRdFi1EUtTvmvA3nG4aNtJdEa1sOtsSAA)

### 💻 Code

* Using `isolatedModules: true`
* In a `.d.ts` file

```ts
declare const enum EventName {
    FOO = 1,
    BAR = 2
}

type EventMap = {
    [EventName.FOO]: number;
    [EventName.BAR]: string;
};
```

### 🙁 Actual behavior

TS fails with the following error:

> Error: [...].d.ts: error TS2748: Cannot access ambient const enums when 'isolatedModules' is enabled.

### 🙂 Expected behavior

TS succeeds

### Workaround

As a workaround, I declare a fake object using the same keys and values as the `enum`, and then use this to declare the `interface`:

```ts
declare const enum EventName {
    FOO = 1,
    BAR = 2
}

declare const EventNameAsConst: {
  FOO: EventName.FOO
  BAR: EventName.BAR
}

type EventMap = {
    [EventNameAsConst.FOO]: number;
    [EventNameAsConst.BAR]: string;
};
```

This works, but it is a bit cumbersome to have to maintain this intermediary object.

### PR DESCRIPTION
**Pull request by a-tarasyuk**
Fixes #54152

### CODE REVIEW

**Comment by gabritto** [comment]
@typescript-bot test this
@typescript-bot run DT
@typescript-bot user test this
@typescript-bot user test tsserver
@typescript-bot test top100
@typescript-bot test tsserver top100
@typescript-bot perf test this

**Comment by gabritto** [comment]
@typescript-bot test this
@typescript-bot run DT
@typescript-bot user test this
@typescript-bot user test tsserver
@typescript-bot test top100
@typescript-bot test tsserver top100
@typescript-bot perf test this

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -37819,7 +37819,7 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
         if (getIsolatedModules(compilerOptions)) {
             Debug.assert(!!(type.symbol.flags & SymbolFlags.ConstEnum));
             const constEnumDeclaration = type.symbol.valueDeclaration as EnumDeclaration;
-            if (constEnumDeclaration.flags & NodeFlags.Ambient) {
+            if (constEnumDeclaration.flags & NodeFlags.Ambient && !(node.flags & NodeFlags.Ambient)) {
```
Comment (gabritto) [inline src/compiler/checker.ts:37822]: I talked to @andrewbranch offline, and we think this change is correct but incomplete, meaning it doesn't capture all of the cases where it is actually safe to use an ambient const enum with `isolatedModules` on.
The idea is that as long as the access isn't going to be emitted, then it is safe and we shouldn't error. This is true if the access occurs in an ambient context, but it is also true in other contexts, e.g. in the original example, even if we're in a normal ts file, it should be ok to use an ambient const enum in a type.
A more complete way to detect when it's safe to access an ambient const enum is to use `isValidTypeOnlyAliasUseSite` from `compiler/utilities.ts`.

**Review by gabritto — CHANGES_REQUESTED** [review]

File: src/compiler/checker.ts (hunk 1 above)
Comment (a-tarasyuk) [reply src/compiler/checker.ts:37822]: @gabritto Thanks for the detailed feedback. I've made the change to use the `isValidTypeOnlyAliasUseSite` helper.

File: tests/cases/compiler/isolatedModulesConstEnum.ts [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+// @filename: /foo.d.ts
+declare const enum EventName {
+    FOO = 1,
+    BAR = 2
+}
+
+type EventMap = {
+    [EventName.FOO]: number;
+    [EventName.BAR]: string;
+};
```
[the remark this answers is not in the document]
Comment (andrewbranch) [reply tests/cases/compiler/isolatedModulesConstEnum.ts:11]:
```suggestion
type EventMap = {
    [EventName.FOO]: number;
    [EventName.BAR]: string;
};

// @filename: bar.ts
type EventMap2 = {
  [EventName.FOO]: number;
  [EventName.BAR]: string;
};

```

**Comment by gabritto** [comment]
@typescript-bot test this
@typescript-bot run DT
@typescript-bot user test this
@typescript-bot user test tsserver
@typescript-bot test top100
@typescript-bot test tsserver top100
@typescript-bot perf test this

**Pull request merged**